streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java

```diff
@@ -0,0 +1,224 @@
+        this.underlying = underlying;
+        this.keySerde = keySerde;
+        this.valueSerde = valueSerde;
+        this.flushListener = flushListener;
+    }
+
+    @Override
+    public String name() {
+        return underlying.name();
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void init(final ProcessorContext context, final StateStore root) {
+        underlying.init(context, root);
+        initInternal(context);
+        // save the stream thread as we only ever want to trigger a flush
+        // when the stream thread is the the current thread.
+        streamThread = Thread.currentThread();
+    }
+
+    @SuppressWarnings("unchecked")
+    void initInternal(final ProcessorContext context) {
+        this.context = (InternalProcessorContext) context;
+        this.serdes = new StateSerdes<>(underlying.name(),
+                                        keySerde == null ? (Serde<K>) context.keySerde() : keySerde,
+                                        valueSerde == null ? (Serde<V>) context.valueSerde() : valueSerde);
+
+        this.name = context.taskId() + "-" + underlying.name();
+        this.cache = context.getCache();
+        cache.addDirtyEntryFlushListener(name, new MemoryLRUCacheBytes.DirtyEntryFlushListener() {
+            @Override
+            public void apply(final List<MemoryLRUCacheBytes.DirtyEntry> entries) {
+                final List<KeyValue<Bytes, byte[]>> keyValues = new ArrayList<>();
+                for (MemoryLRUCacheBytes.DirtyEntry entry : entries) {
+                    keyValues.add(KeyValue.pair(entry.key(), entry.newValue()));
+                    flushListener.forward(serdes.keyFrom(entry.key().get()),
+                                          change(entry.newValue(), underlying.get(entry.key())),
+                                          entry.recordContext(),
+                                          CachingKeyValueStore.this.context);
+                }
+                underlying.putAll(keyValues);
+                underlying.flush();
+            }
+        });
+
+    }
+
+    @Override
+    public synchronized void flush() {
+        cache.flush(name);
+    }
+
+    private Change<V> change(final byte[] newValue, final byte[] oldValue) {
+        if (sendOldValues) {
+            return new Change<>(serdes.valueFrom(newValue), serdes.valueFrom(oldValue));
+        }
+        return new Change<>(serdes.valueFrom(newValue), null);
+    }
+
+    @Override
+    public void close() {
+        flush();
+        underlying.close();
+    }
+
+    @Override
+    public boolean persistent() {
+        return underlying.persistent();
+    }
+
+    @Override
+    public boolean isOpen() {
+        return underlying.isOpen();
+    }
+
+    @Override
+    public void enableSendingOldValues() {
+        this.sendOldValues = true;
+    }
+
+    @Override
+    public synchronized V get(final K key) {
```

Comment:
Just thinking if `synchronized` here will have large performance impact; and thus, if it would be better to have a cache per thread instead of a JVM global cache?